# IMPORT DATA

In [1]:
from google.colab import drive
drive.mount('/content/Drive')

Mounted at /content/Drive


In [2]:
import pandas as pd
import numpy as np

In [3]:
train_df = pd.read_csv('/content/Drive/MyDrive/House_price_predictor/dataset/featured_data/featured_train_data.csv')
test_df = pd.read_csv('/content/Drive/MyDrive/House_price_predictor/dataset/featured_data/featured_test_data.csv')

In [4]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

train_df.head()

median_sale_price  median_list_price  median_ppsf  median_list_ppsf  \
0            46550.0           217450.0    31.813674        110.183666   
1            61870.0           245000.0    40.723982        130.528256   
2           125500.0           217450.0    63.913043        119.919216   
3           153000.0           189900.0    81.598080        105.617353   
4           165500.0           154000.0    81.598080         83.921175   

   homes_sold  pending_sales  inventory  median_dom  avg_sale_to_list  \
0        14.0           23.0       64.0        59.5          0.943662   
1        22.0           29.0       69.0        89.5          0.946642   
2        24.0           40.0       60.0       144.5          0.955624   
3        34.0           46.0       57.0       126.0          0.970608   
4        39.0           49.0       50.0        80.0          0.982105   

   sold_above_list  off_market_in_two_weeks  year  bank  bus  hospital  mall  \
0         0.142857                 0.043478  2012  12.0  2.0       4.0   1.0   
1         0.090909                 0.034483  2012  12.0  2.0       4.0   1.0   
2         0.208333                 0.100000  2012  12.0  2.0       4.0   1.0   
3         0.176471                 0.108696  2012  12.0  2.0       4.0   1.0   
4         0.256410                 0.102041  2012  12.0  2.0       3.0   1.0   

   park  restaurant  school  station  supermarket  Median_Age  \
0  60.0        45.0    57.0      4.0          7.0        36.3   
1  60.0        45.0    57.0      4.0          7.0        36.3   
2  60.0        45.0    57.0      4.0          7.0        36.3   
3  60.0        45.0    57.0      4.0          7.0        36.3   
4  60.0        45.0    57.0      4.0          6.0        36.3   

   Per_Capita_Income  Median_Rent  Median_Home_Value  Unemployed_Population  \
0            33052.0        710.0           279500.0                  460.0   
1            33052.0        710.0           279500.0                  460.0   
2            33052.0        710.0           279500.0                  460.0   
3            33052.0        710.0           279500.0                  460.0   
4            33052.0        710.0           279500.0                  460.0   

           price  Latitude  Longitude  month  quarter  zipcode_freq  \
0  200773.999557     33.75     -84.39      3        1            94   
1  202421.064584     33.75     -84.39      4        2            94   
2  202681.309539     33.75     -84.39      5        2            94   
3  202998.603897     33.75     -84.39      6        2            94   
4  203781.903446     33.75     -84.39      7        3            94   

          cityTE  
0  192580.125925  
1  192580.125925  
2  192580.125925  
3  192580.125925  
4  192580.125925

# SPLIT TRAIN AND TEST DATA

In [5]:
X_train = train_df.drop('price', axis=1)
y_train = train_df['price']

X_test = test_df.drop('price', axis=1)
y_test = test_df['price']

In [6]:
X_train.columns

Index(['median_sale_price', 'median_list_price', 'median_ppsf',
       'median_list_ppsf', 'homes_sold', 'pending_sales', 'inventory',
       'median_dom', 'avg_sale_to_list', 'sold_above_list',
       'off_market_in_two_weeks', 'year', 'bank', 'bus', 'hospital', 'mall',
       'park', 'restaurant', 'school', 'station', 'supermarket', 'Median_Age',
       'Per_Capita_Income', 'Median_Rent', 'Median_Home_Value',
       'Unemployed_Population', 'Latitude', 'Longitude', 'month', 'quarter',
       'zipcode_freq', 'cityTE'],
      dtype='object')

# BASEMODEL -> DUMMY REGRESSOR

In [7]:
'''
import mlflow.sklearn
import mlflow.sklearn
'''
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from sklearn.dummy import DummyRegressor

model = DummyRegressor(strategy='median')
model.fit(X_train, y_train)

y_pred = model.predict(X_test)


mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2_score = r2_score(y_test, y_pred)

print('Model -> Dummy Regressor')
print(f'Mean absolute error -> {mae}')
print(f'Mean squared error -> {mse}')
print(f'R2 score -> {r2_score}')

'''
mlflow.set_experiment("DummyRegressor")
mlflow.set_tracking_uri("http://localhost:5000")
with mlflow.start_run():
  mlflow.log_param('model', model.strategy)
  mlflow.log_metric("mean_absolute_error", mae)
  mlflow.log_metric("mean_squared_error", mse)
  mlflow.log_metric("r2_score", r2_score)

  mlflow.sklearn.log_model(model , artifact_path='Dummy_Regressor_model')
  print('Logged successfully')
  '''

Model -> Dummy Regressor
Mean absolute error -> 267098.01570576796
Mean squared error -> 200735998946.33554
R2 score -> -0.35405980689716166


'\nmlflow.set_experiment("DummyRegressor")\nmlflow.set_tracking_uri("http://localhost:5000")\nwith mlflow.start_run():\n  mlflow.log_param(\'model\', model.strategy)\n  mlflow.log_metric("mean_absolute_error", mae)\n  mlflow.log_metric("mean_squared_error", mse)\n  mlflow.log_metric("r2_score", r2_score)\n\n  mlflow.sklearn.log_model(model , artifact_path=\'Dummy_Regressor_model\')\n  print(\'Logged successfully\')\n  '

 As we can see from above, this model performed really bad and cannot be used.

# LINEAR REGRESSION

In [8]:
'''
import mlflow.sklearn
import mlflow.sklearn
'''
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

SS = StandardScaler()
X_train_scaled = SS.fit_transform(X_train)
X_test_scaled = SS.transform(X_test)

y_train_log = np.log1p(y_train)
y_test_log = np.log1p(y_test)

In [9]:
LR = LinearRegression()
RD = Ridge()
LS = Lasso()
EN = ElasticNet()

models = [LR, RD, LS, EN]

for model in models:
  model.fit(X_train_scaled, y_train_log)
  y_pred_log = model.predict(X_test_scaled)
  y_pred = np.expm1(y_pred_log)


  print(str(model))
  print(f'mean absolute error -> {mean_absolute_error(y_test, y_pred)}')
  print(f'mean squared error -> {mean_squared_error(y_test, y_pred)}')
  print(f'r2 score -> {r2_score(y_test, y_pred)}')
  print("")
  print("----------------------")

LinearRegression()
mean absolute error -> 296248.62018313626
mean squared error -> 51120526683901.61
r2 score -> -343.8322715079633

----------------------
Ridge()
mean absolute error -> 296249.4745659361
mean squared error -> 51123089826933.46
r2 score -> -343.8495611270499

----------------------
Lasso()
mean absolute error -> 267941.8302127299
mean squared error -> 201495895264.34488
r2 score -> -0.35918566905953675

----------------------
ElasticNet()
mean absolute error -> 258898.31104118383
mean squared error -> 186091183770.77222
r2 score -> -0.2552735617156523

----------------------


# XGBREGRESSOR

In [10]:
from xgboost import XGBRegressor

In [12]:
model_xgb = XGBRegressor()
model.fit(X_train, y_train_log)
y_pred = model.predict(X_test)
y_pred = np.expm1(y_pred)

print(f'XGBRegressor model')
print(f'mean absolute error -> {mean_absolute_error(y_test, y_pred)}')
print(f'mean squared error -> {mean_squared_error(y_test, y_pred)}')
print(f'r2 score -> {r2_score(y_test, y_pred)}')
print("")
print("----------------------")

XGBRegressor model
mean absolute error -> 41548.55584821438
mean squared error -> 9011144945.622602
r2 score -> 0.9392155405655264

----------------------


## XGBoost performed the best overall. It required less time and fewer computational resources, and it achieved better performance than TensorFlow. The TensorFlow model recorded an MAE of 53,935.74, whereas the XGBoost model achieved a lower MAE of 41,548.56, indicating superior predictive accuracy.

# TENSORFLOW MODEL

In [ ]:
X_train_scaled.shape , X_test_scaled.shape

((578815, 32), (297631, 32))

In [ ]:
import tensorflow as tf
from tensorflow.keras import Sequential, Input
from keras.layers import Dense
from tensorflow.keras.callbacks import EarlyStopping

SS = StandardScaler()
X_train_scaled = SS.fit_transform(X_train)
X_test_scaled = SS.transform(X_test)

y_train_log = np.log1p(y_train)
y_test_log = np.log1p(y_test)


model = Sequential([
    Input(shape=(32,)),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(1)
])
model.compile(optimizer='adam', loss='mse', metrics=['mae'])

from sklearn.model_selection import train_test_split
X_tr , X_val, y_tr, y_val = train_test_split(X_train_scaled, y_train_log, test_size=0.2, random_state=42)

early_stopping = EarlyStopping(monitor='val_mae', patience=5, restore_best_weights=True)


model.fit(X_tr, y_tr,
          epochs=300,
          validation_data=(X_val, y_val),
          callbacks=[early_stopping])

y_pred = model.predict(X_test_scaled)
y_pred = np.expm1(y_pred)
y_pred = y_pred.ravel()

mae = mean_absolute_error(y_test, y_pred)

print('validation mae : ', mae)

Epoch 1/300
14471/14471 ━━━━━━━━━━━━━━━━━━━━ 73s 5ms/step - loss: 3.4292 - mae: 0.4945 - val_loss: 0.0591 - val_mae: 0.1294
Epoch 2/300
14471/14471 ━━━━━━━━━━━━━━━━━━━━ 59s 4ms/step - loss: 0.0478 - mae: 0.1242 - val_loss: 0.0240 - val_mae: 0.1003
Epoch 3/300
14471/14471 ━━━━━━━━━━━━━━━━━━━━ 54s 4ms/step - loss: 0.0277 - mae: 0.1080 - val_loss: 0.0187 - val_mae: 0.0920
Epoch 4/300
14471/14471 ━━━━━━━━━━━━━━━━━━━━ 55s 4ms/step - loss: 0.0337 - mae: 0.1000 - val_loss: 0.0157 - val_mae: 0.0825
Epoch 5/300
14471/14471 ━━━━━━━━━━━━━━━━━━━━ 52s 4ms/step - loss: 0.0248 - mae: 0.0941 - val_loss: 0.0148 - val_mae: 0.0821
Epoch 6/300
14471/14471 ━━━━━━━━━━━━━━━━━━━━ 53s 4ms/step - loss: 0.0225 - mae: 0.0915 - val_loss: 0.0209 - val_mae: 0.1109
Epoch 7/300
14471/14471 ━━━━━━━━━━━━━━━━━━━━ 54s 4ms/step - loss: 0.0206 - mae: 0.0891 - val_loss: 0.0160 - val_mae: 0.0753
Epoch 8/300
14471/14471 ━━━━━━━━━━━━━━━━━━━━ 56s 4ms/step - loss: 0.0166 - mae: 0.0857 - val_loss: 0.0167 - val_mae: 0.0821
Epoch 9/

# MODEL ANALYSIS

XGBoost performed the best overall. It required less time and fewer computational resources, and it achieved better performance than TensorFlow.
The TensorFlow model recorded an MAE of 53,935.74, whereas the XGBoost model achieved a lower MAE of 41,548.56, indicating superior predictive accuracy.

# MODEL SAVE

In [15]:
import pickle

model_path = '/content/Drive/MyDrive/House_price_predictor/model_xgb.pkl'

with open(model_path , 'wb') as f:
  pickle.dump(model_xgb , f)